File: Project/QuizTime/app/src/main/java/com/mrahmed/quiztime/Adapter/QuestionAdapter.java
package com.mrahmed.quiztime.Adapter;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import androidx.recyclerview.widget.AsyncListDiffer;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;
import com.mrahmed.quiztime.Domain.QuestionModel;
import com.mrahmed.quiztime.Domain.UserModel;
import com.mrahmed.quiztime.databinding.ViewholderQuestionBinding;
import androidx.core.content.ContextCompat;
import android.graphics.Color;

import java.util.ArrayList;
import java.util.List;

public class QuestionAdapter extends RecyclerView.Adapter<QuestionAdapter.ViewHolder> {

    private String correctAnswer;
    private List<String> users; // Change MutableList to List
    private Score returnScore;
    private AsyncListDiffer<String> differ;

    public interface Score {
        void amount(int number, String clickedAnswer);
    }

    public QuestionAdapter(String correctAnswer, List<String> users, Score returnScore) {
        this.correctAnswer = correctAnswer;
        this.users = (users != null) ? users : new ArrayList<>(); // Use ArrayList instead of MutableList
        this.returnScore = returnScore;

        differ = new AsyncListDiffer<>(this, new DiffUtil.ItemCallback<String>() {
            @Override
            public boolean areContentsTheSame(String oldItem, String newItem) {
                return oldItem.equals(newItem);
            }

            @Override
            public boolean areItemsTheSame(String oldItem, String newItem) {
                return oldItem.equals(newItem);
            }
        });
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        ViewholderQuestionBinding binding;

        public ViewHolder(ViewholderQuestionBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        ViewholderQuestionBinding binding = ViewholderQuestionBinding.inflate(inflater, parent, false);
        return new ViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        String answer = differ.getCurrentList().get(position);
        holder.binding.answerTxt.setText(answer);

        int currentPos = getCurrentPosition();

        // Check if the answer is correct
        if (differ.getCurrentList().size() == 5 && currentPos == position) {
            holder.binding.answerTxt.setBackgroundResource(R.drawable.green_background);
            holder.binding.answerTxt.setTextColor(Color.WHITE);
            holder.binding.answerTxt.setCompoundDrawablesRelativeWithIntrinsicBounds(
                    null, null, ContextCompat.getDrawable(holder.itemView.getContext(), R.drawable.tick), null);
        }

        // Handle clicked answer
        if (differ.getCurrentList().size() == 5) {
            int clickedPos = getClickedPosition();
            if (clickedPos == position && clickedPos != currentPos) {
                holder.binding.answerTxt.setBackgroundResource(R.drawable.red_background);
                holder.binding.answerTxt.setTextColor(Color.WHITE);
                holder.binding.answerTxt.setCompoundDrawablesRelativeWithIntrinsicBounds(
                        null, null, ContextCompat.getDrawable(holder.itemView.getContext(), R.drawable.thievea), null);
            }
        }

        // Hide the 5th position
        if (position == 4) {
            holder.itemView.setVisibility(View.GONE);
        }

        // Handle click event
        holder.itemView.setOnClickListener(v -> {
            String str = getAnswerString(position);
            users.add(str); // Add clicked answer to the users list
            notifyDataSetChanged();

            if (currentPos == position) {
                holder.binding.answerTxt.setBackgroundResource(R.drawable.green_background);
                holder.binding.answerTxt.setTextColor(Color.WHITE);
                holder.binding.answerTxt.setCompoundDrawablesRelativeWithIntrinsicBounds(
                        null, null, ContextCompat.getDrawable(holder.itemView.getContext(), R.drawable.tick), null);
            } else {
                holder.binding.answerTxt.setBackgroundResource(R.drawable.red_background);
                holder.binding.answerTxt.setTextColor(Color.WHITE);
                holder.binding.answerTxt.setCompoundDrawablesRelativeWithIntrinsicBounds(
                        null, null, ContextCompat.getDrawable(holder.itemView.getContext(), R.drawable.thievea), null);
                returnScore.amount(0, str);
            }
        });
    }

    @Override
    public int getItemCount() {
        return differ.getCurrentList().size();
    }

    private int getCurrentPosition() {
        // Logic to get the current position based on correctAnswer
        switch (correctAnswer) {
            case "a":
                return 0;
            case "b":
                return 1;
            case "c":
                return 2;
            case "d":
                return 3;
            default:
                return -1;
        }
    }

    private int getClickedPosition() {
        // Logic to get the clicked position based on the last answer clicked
        return users.size() > 0 ? getCurrentPosition(users.get(users.size() - 1)) : -1;
    }

    private String getAnswerString(int position) {
        switch (position) {
            case 0:
                return "a";
            case 1:
                return "b";
            case 2:
                return "c";
            case 3:
                return "d";
            default:
                return "";
        }
    }
}

File: Project/QuizTime/app/src/main/java/com/mrahmed/quiztime/Adapter/QuestionAdapter.kt
package com.mrahmed.quiztime.Adapter

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.mrahmed.quiztime.Domain.QuestionModel
import com.mrahmed.quiztime.Domain.UserModel
import com.mrahmed.quiztime.databinding.ViewholderQuestionBinding

class QuestionAdapter(
    val correctAnswer: String,
    val users: MutableList<String> = mutableListOf(),
    val returnScore: score

) : RecyclerView.Adapter<QuestionAdapter.Viewholder>() {

    interface score {
        fun amount(number: Int, clickedAnswer: String)
    }

    class Viewholder {}

    private lateinit var binding: ViewholderQuestionBinding
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): QuestionAdapter.Viewholder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = ViewholderQuestionBinding.inflate(inflater, parent, false)
        return Viewholder()
    }

    override fun onBindViewHolder(holder: QuestionAdapter.Viewholder, position: Int) {
        val binding = ViewholderQuestionBinding.bind(holder.itemView)
        binding.answerTxt.text = differ.currentList[position]
        var currentPos = 0
        when (currentAnswer) {
            "a" -> {
                currentPos = 0
            }

            "b" -> {
                currentPos = 1
            }

            "c" -> {
                currentPos = 2
            }

            "d" -> {
                currentPos = 3
            }
        }
        if (differ.currentList.size == 5 && currentPos == position) {
            binding.answerTxt.setBackgroundResource(R.drawable.green_background)
            binding.answerTxt.setTextColor(
                ContextCompact.getCcolor(
                    binding.root.context,
                    R.color.white
                )
            )

            val drawable = ContextCompact.getDrawable(binding.root.context, R.drawable.tick)
            binding.answerTxt.setCompoundDrawablesRelativeWithIntrinsicBounds(
                null,
                null,
                drawable,
                null
            )

        }
        if (differ.currentList.size == 5) {
            var clickedPos = 0
            when (differ.currentAnswer[4]) {
                "a" -> {
                    clickedPos = 0
                }

                "b" -> {
                    clickedPos = 1
                }

                "c" -> {
                    clickedPos = 2
                }

                "d" -> {
                    clickedPos = 3
                }
            }
            if (clickedPos == position && clickedPos != currentPos) {
                binding.answerTxt.setBackgroundResource(R.drawable.red_background)
                binding.answerTxt.setTextColor(
                    ContextCompact.getCcolor(
                        binding.root.context,
                        R.color.white
                    )
                )

                val drawable = ContextCompact.getDrawable(binding.root.context, R.drawable.thievea)
                binding.answerTxt.setCompoundDrawablesRelativeWithIntrinsicBounds(
                    null,
                    null,
                    drawable,
                    null
                )
            }

        }
        if (position == 4) {
            binding.root.visibility= View.GONE
        }
        holder.itemView.setOnClickListener {
            var str=""
            when (position) {
                0 -> {
                    str = "a"
                }
                1 -> {
                    str = "b"
                }
                2 -> {
                    str = "c"
                }
                3 -> {
                    str = "d"
                }
            }
            users.add(4, str){}
        notifyDataSetChanged()
            if (correctPos == position){
                binding.answerTxt.setBackgroundResource(R.drawable.green_background)
                binding.answerTxt.setTextColor(
                    ContextCompact.getCcolor(
                        binding.root.context,
                        R.color.white
                    )
                )

                val drawable = ContextCompact.getDrawable(binding.root.context, R.drawable.tick)
                binding.answerTxt.setCompoundDrawablesRelativeWithIntrinsicBounds(
                    null,
                    null,
                    drawable,
                    null
                )
            }
            else{
                binding.answerTxt.setBackgroundResource(R.drawable.red_background)
                binding.answerTxt.setTextColor(
                    ContextCompact.getCcolor(
                        binding.root.context,
                        R.color.white
                    )
                )

                val drawable = ContextCompact.getDrawable(binding.root.context, R.drawable.thievea)
                binding.answerTxt.setCompoundDrawablesRelativeWithIntrinsicBounds(
                    null,
                    null,
                    drawable,
                    null
                )
                returnScore.amount(0,str)
            }
        }
        if (differ.currentList.size=5) holder.itemView.setOnClickListener (null)
    }

    override fun getItemCount(): differ.currentList.size

    inner class ViewHolder : RecyclerView.ViewHolder(binding.root)

    private val differCallback = object : DiffUtil.ItemCallback<String>() {
        override fun areContentsTheSame(oldItem: String, newItem: String): Boolean {
            return oldItem == newItem
        }

        override fun areItemsTheSame(oldItem: String, newItem: String): Boolean {
            return oldItem == newItem
        }

    }

    val differ = AsyncListDiffer(this, differCallback)


}